provider/server.cpp

```diff
@@ -233,24 +239,370 @@ handle_return (
 }
 
 
+void
+close_listener_socket (
+    int* fd)
+{
+    SCX_BOOKEND ("close_listener_socket");
+    close (*fd);
+}
+
+
+void
+generate_key (
+    unsigned int (&key)[4])
+{
+    srand (time (0));
+    for (int i = 0; i < 4; ++i)
+    {
+        key[i] = 0;
+        for (int j = 0; j < 4; ++j)
+        {
+            key[i] |= static_cast<unsigned int>(
+                static_cast<unsigned char> (rand () % 0xff) << (8 * j));
+        }
+    }
+}
+
+
+int
+create_listener (
+    unsigned short* pPortOut,
+    int* pListenerFDOut)
+{
+    SCX_BOOKEND ("create_listener");
+    assert (pPortOut);
+    assert (pListenerFDOut);
+    int result = Server::LISTEN_SOCKET_FAILED;
+    util::unique_ptr<int, void (*)(int*)> listener_holder (NULL, close_listener_socket);
+    int listener_fd = socket (AF_INET, SOCK_STREAM, 0);
+    if (-1 != listener_fd)
+    {
+        listener_holder.reset (&listener_fd);
+        int flags = fcntl (listener_fd, F_GETFL, 0);
+        if (-1 != flags &&
+            -1 != fcntl (listener_fd, F_SETFL, flags | O_NONBLOCK))
+        {
+            sockaddr_in addr;
+            memset (&addr, 0, sizeof (addr));
+            addr.sin_family = AF_INET;
+            addr.sin_addr.s_addr = INADDR_ANY;
+            unsigned short port = 5999;
+            int local_result;
+            do
+            {
+                ++port;
+                addr.sin_port = htons (port);
+                local_result =
+                    bind (listener_fd, reinterpret_cast<sockaddr*>(&addr),
+                          sizeof (addr));
+            } while (-1 == local_result &&
+                     EADDRINUSE == errno);
+            if (0 == local_result)
+            {
+                local_result = listen (listener_fd, 5);
+                if (0 == local_result)
+                {
+                    *pPortOut = port;
+                    *pListenerFDOut = listener_fd;
+                    listener_holder.release ();
+                    result = Server::SUCCESS;
+                }
+#if (PRINT_BOOKENDS)
+                else
```

Comment:
need to set result for other 3 else condition?

Reply:
Result is initialized to a fail condition.  The case where it is set is the single success case.  This is as designed.